plugins/nominal-connection-checker/src/type_hierarchy.js

```diff
@@ -439,4 +563,140 @@ class TypeDef {
   hasDescendant(descendantName) {
     return this.descendants_.has(descendantName);
   }
+
+  /**
+   * Returns an array of this type's parameters, in the order for its superType.
+   * @param {string} ancestorName The caseless name of the ancestor to get the
+   *     parameters for.
+   * @param {!Array<!TypeStructure>=} actualTypes Optional actual types to
+   *     substitute for parameters. These types may be generic.
+   * @return {!Array<!TypeStructure>} This type's parameters, in the order for
+   *     its superType.
+   */
+  getParamsForAncestor(ancestorName, actualTypes = undefined) {
+    if (ancestorName == this.name && !this.paramsMap_.has(this.name)) {
+      // Convert this type's params to a type structure.
+      this.paramsMap_.set(
+          this.name,
+          this.params_.map((param) => {
+            return {name: param.name, params: []};
+          }));
+    }
+    const params = this.paramsMap_.get(ancestorName);
+    if (actualTypes) {
+      const replaceFn = (param, i, array) => {
+        const paramIndex = this.getIndexOfParam(param.name);
+        if (paramIndex != -1) {
+          array[i] = actualTypes[paramIndex];
+        } else {
+          param.params.forEach(replaceFn, this);
+        }
+      };
+      params.forEach(replaceFn, this);
+    }
+    return params;
+  }
+
+  /**
+   * Returns true if this type has any parameters. False otherwise.
+   * @return {boolean} True if this type has any parameters. False otherwise.
+   */
+  hasParameters() {
+    return !!this.params_.length;
+  }
+
+  /**
+   * Returns true if this type has a parameter with the given name.
+   * False otherwise.
+   * @param {string} paramName The caseless name of the possible parameter.
+   * @return {boolean} True if this type has a parameter with the given name.
+   *     False otherwise.
+   */
+  hasParameter(paramName) {
+    return this.params_.some((param) => param.name == paramName);
+  }
+
+  /**
+   * Returns the index of the parameter with the given name, or -1 if the
+   * parameter does not exist..
+   * @param {string} paramName The name of the parameter.
+   * @return {number} The index of hte parameter.
+   */
+  getIndexOfParam(paramName) {
+    return this.params_.findIndex((param) => param.name == paramName);
+  }
+
+  /**
+   * Returns the parameter definition for the parameter at the given index.
+   * @param {number} index The index to get the parameter definition of.
+   * @return {!ParamDef} The parameter definition for the parameter at the
+   *     given index.
+   */
+  getParamForIndex(index) {
+    return this.params_[index];
+  }
+
+  /**
+   * Returns the ParamDef with the given name, or undefined if not found.
+   * @param {string} paramName The name of the parameter to find the
+   *     ParamDef of.
```

Comment:
We're trying to reduce lint so that we can turn on warnings as errors for pull requests--please fix this one by moving `the` down a line.